### ISSUE TITLE
Let redundancy groups not just fail

### ISSUE THREAD

**Issue by nilmerg** [Icinga/icinga2#10190]
This is not a bug, not yet, at least. Once #10014 is fixed, this should be considered.

## tl/dr

Redundancy groups should not just fail, they should in addition be *not reachable* or *not determinable*, just like hosts and services are.

## What's this about?

```mermaid
graph LR;
    ChildHost-->pa["ParentHostA (Group 1)"];
    ChildHost-->pb["ParentHostB (Group 1)"];
    pa-->ga["GrandParentA"];
    pb-->gb["GrandParentB"];
```

Right now, the child host will be unreachable in two cases:
1. Both parents are down (expected)
2. One of them is unreachable (The linked bug)

Once the second case is fixed, it will likely behave like this instead:
1. Both parents are down (still, expected)
2. Both parents are unreachable (new, expected)

But what happens in case only one of the parents is down and the other unreachable?

3. One parent is down and the other unreachable (new, ??)

I suppose so, since already right now, the availability of the child host is influenced by its parent's reachable state. It will just be extended to consider redundancy groups in such a case.

## But then there is `Dependency::IsAvailable`

Which is what is used to determine a redundancy group's state at the moment. Though, it doesn't consider the parent's reachable state at all. The parent might be UP, but unreachable, and it returns true.

So, in order to really ensure that the child host is unreachable in the third case, this might need to change. (The potential bug)

Of course, this has larger side-effects, as it changes how dependencies work in general.

Which is what this issue is really about! As I believe this is a good thing. Why shouldn't a dependency fail, in case another one higher up in the hierarchy fails? This is already the case, I think, anyway. This discrepancy will only pop up in case checks are disabled on the parent in question. (As if checks run, the parent will eventually go down/unknown)

## Additional note

What led me to this, was the wish to understand how a redundancy group's state will be determined as part of https://github.com/Icinga/icingadb/issues/347.

As of today I thought it simply represents whether all related dependencies failed or not. But now a *failed* dependency might be failing, because a parent might just not be reachable. And so will a redundancy group.

This all sounds fine, unless you want to identify the actual root problem in a failing dependency chain. (A root problem is a dependency node, which is reachable and has a problem)

In a dependency chain, the current plan is, to represent redundancy groups as actual nodes with an actual state. While hosts and services also have state, they can in addition be not reachable. Redundancy groups can only fail.

But a redundancy group, with one host member that is UP but not reachable, is technically also not reachable. So it's not just a failing dependency, but one that is … not determinable?

**Comment by yhabteab** [comment]
> Which is what this issue is really about!

To be honest, I don't quite understand what this issue is all about :). You have listed a number of open questions that don't quite fit the title of the issue.

> Redundancy groups should not just fail, they should in addition be _not reachable_ or _not determinable_, just like hosts and services are.

This particular case might be covered by #10097 as that issue is about syncing dependencies to Redis.

> But what happens in case only one of the parents is down and the other unreachable?
>
> Which is what is used to determine a redundancy group's state at the moment. Though, it doesn't consider the parent's reachable state at all. The parent might be UP, but unreachable, and it returns true.
>
> So, in order to really ensure that the child host is unreachable in the third case, this might need to change. (The potential bug)

The `Dependency::IsAvailable()` method is concerned with verifying the availability of a given dependency, rather than determining the reachability of its direct or intermediate parents. With #10228, the child host will of course<sub> at least for me</sub> be unreachable. It is not entirely clear to me, however, why this particular case should be treated differently from the others.

> I suppose so, since already right now, the availability of the child host is influenced by its parent's reachable state. It will just be extended to consider redundancy groups in such a case.

Isn't that what #10014 is about? That's a confirmed bug that will be fixed by #10228, but I don't see how that applies to this issue.

> Why shouldn't a dependency fail, in case another one higher up in the hierarchy fails? This is already the case, I think, anyway.

This is already the case, all parents in a given dependency graph are evaluated recursively, but no redundancy groups have been taken into account so far.


> But a redundancy group, with one host member that is UP but not reachable, is technically also not reachable. So it's not just a failing dependency, but one that is … not determinable?

Well, it's simply unreachable. Isn't the _unreachable_ state intended for use cases like this? There's no need to add another meaning to it.

**Comment by nilmerg** [comment]
Recently, we discussed this in person and I wasn't so sure anymore whether we required `failed` and `is_reachable`, or just one of them. Now I'am sure, we need both. Let me explain:

Note that, as always, implicit relations (host -> service) are of no importance here.

 ```mermaid
 graph TD;
     A-->B;
     B-->D;
     C-->D;
     subgraph BC [ redundancy group ]
     B~~~C
    end
 ```

The group is *failed* if `B` and `C` are *reachable* and cause `D` to be not *reachable*.
The group is not *reachable* if either `B` or `C` are as well not *reachable*. In this example only `B` cannot.

#10158 talks of `affects_children` for host and service states. This boolean is very similar to my definition of a group's `failed` state. Both are only true, if the state can be reliably determined and is responsible for any children to be not *reachable*. (i.e. the object in question is *reachable*)

If `affects_children` or `failed` is true, `is_reachable` can by definition then also only be true. On the other hand, if `is_reachable` is false, `affects_children` and `failed` cannot be determined and hence cannot be relied upon.

Now, the root problem of say `D`, is the closest parent that is *reachable* and responsible. (`affects_children=true`, `failed=true`)

 ```mermaid
 graph TD;
     A[down]-->B;
     B["up (unreachable)"]-->D;
     C[down]-->D["down (unreachable)"];
     subgraph BC [ unreachable ]
     B~~~C
    end
 ```

In this case `D` is not *reachable* and `A` is clearly the root problem.
